File: Sorting/Merge_Sort.java
package com.sort;
import java.util.*;

public class MergeSort{
    public static void main(String[] args) {
        int[] arr = {11,30,24,7,31,16,39,41};
        System.out.println("Array before sorting: ");
        System.out.println(Arrays.toString(arr));
        mergeSort(arr,0,arr.length-1);
        System.out.println("Array after sorting: ");
        System.out.println(Arrays.toString(arr));
    }
    static void mergeSort(int[] arr,int lb,int ub) {
        if (lb < ub) {
            int mid = (lb + ub) / 2;
            mergeSort(arr, lb, mid);
            mergeSort(arr, mid + 1, ub);
            merge(arr, lb, mid, ub);
        }
    }
    static void merge(int[] arr, int lb, int mid, int ub)
    {
        int i = lb, j = mid+1,k = lb;
        int[] b = new int[arr.length];

        while(i <= mid && j <= ub) {
            if(arr[i] <= arr[j]) {
                b[k] = arr[i];
                i++;
            }
            else {
                b[k] = arr[j];
                j++;
            }
            k++;
        }
        if(i > mid){
            while(j <= ub){
                b[k] = arr[j];
                k++;j++;
            }
        }
        else{
            while(i <= mid){
                b[k] = arr[i];
                k++;i++;
            }
        }
        for(k = lb; k <= ub; k++)
            arr[k] = b[k];
    }
}
/*
----------------- Output --------------------

Array before sorting: 
[11, 30, 24, 7, 31, 16, 39, 41]
Array after sorting: 
[7, 11, 16, 24, 30, 31, 39, 41]

*/

File: Sorting/Merge_Sort.py
"""
Merge Sort is one of the most popular sorting algorithms that is based on the principle of Divide and Conquer Algorithm.
Here, a problem is divided into multiple sub-problems. Each sub-problem is solved individually.
 Finally, sub-problems are combined to form the final solution.
 
 Time Complexity : 
          Best Case Complexity   :     O(n*log n)
          Worst Case Complexity  :     O(n*log n)
          Average Case Complexity:     O(n*log n)
Space Complexity : 
          The space complexity of merge sort is O(n).
"""

def mergeSort(arr):
    if len(arr) > 1:
        mid = len(arr)//2
        L = arr[:mid]
        R = arr[mid:]

        mergeSort(L)
        mergeSort(R)

        i = j = k = 0
        while i < len(L) and j < len(R):
            if L[i] < R[j]:
                arr[k] = L[i]
                i += 1
            else:
                arr[k] = R[j]
                j += 1
            k += 1

        while i < len(L):
            arr[k] = L[i]
            i += 1
            k += 1

        while j < len(R):
            arr[k] = R[j]
            j += 1
            k += 1

test_cases = [
    [78,-90,43,0,45,2,-1000,143],
    [],
    [3],
    [9, 8, 7, 2],
    [1, 2, 3, 4, 5]
]

for arr in test_cases:
    mergeSort(arr)
    print(arr)

    
    
